### DISCUSSION TITLE
3d collision in 2d isometric map in Monogame

### DISCUSSION THREAD

**Discussion by BilalAli2008** [MonoGame/MonoGame#8592] [Desktop computers]
I made a map in monogame, An isometric map. After making a map i tried to ad 3d collision in the 2d isometric map but nothing worked. I tried the rectangle intersect method, the diamond method, the bounding box method, but nothing worked.....,
![30 11 2024_18 39 30_REC](https://github.com/user-attachments/assets/7b3562e4-9899-4653-a2d4-20fb95f8f2b4)
If you see the photo you would see a tile (which is the player in a ground tile), that is surrounded by red tiles. The red tiles shows the collision, as you can see it is not perfect. I am using Vector3 and the z position is the depth which for the player is 1 and the ground (not including trees and wood) is also 1. Currently i am the bounding box  method. It could be that i am using the method wrong or the formulas wrong

**Comment by trinith** [comment]
Without knowing more about your code and implementation, it's hard to give any advice. I know the presentation of your map is 2D isometric, but is the data itself either 3D, or layered 2D tilesets?

I assume you would want your player to only collide with the trunk of the tree but otherwise be free to move underneath the braches of the tree. If so, you should be definitely consider how your data is arranged. In this context I would expect you to be querying a 2D layer of some kind for the layers your player character intersects with to get the tiles that they could intersect with. This wouldn't be much different than normal 2D collision, except that you just get data from the multiple layers.

Anyway, as you can see, there's a lot of speculation in my post here. I dunno how helpful that is but without knowing more about your goals, data, and implementation, I'd say maybe just focus on 2D collision first and then build on that.

**Reply by BilalAli2008** [reply]
So, the tile maps are 2d, and the the data to build the map is from different layers (layer 1,2,3,4,5). The tile map itself is 2d however i added Vector3 and Vector3.z is the depth. I thought using Vector3 i would be able to make 3d collision.

**Reply by trinith** [reply]
If you've already got things organized into layers, that's great! I don't think you can just add a Vector3 without doing anything else though since that doesn't really take into account the data you have.

Each of your layers seems to represent a 1x1x1 world unit of space meaning the height for each layer is 1. This means you can just check each layer for the area of space the object your'e checking occupies. In the easiest scenario, and that seemingly demonstrated by your picture, your object is a height of one. This means you only have to check one layer. Assuming your object's Z position corresponds to the index of the layer its on, you can just query your array of layers to get the layer the object is on and do a standard 2D collision check for that layer.

If your object's height is larger than 1, then start at the first layer and go up each layer in turn. For example...

```
for (int i = player.Position.Z; i < layers.Count && i < player.Position.Z + player.Size.Z; i++)
{
  // Check if player collides with any tiles on layers[i].
  // If the player collides and you don't need any additional collision data, you can break out of this loop since you've already collided.
}
```

Since your object only seems to be one layer high though, just start with the base case and move out from there. Perform the check on the layer the player is currently, just like you would for any single layer 2D tile collision.

**Reply by BilalAli2008** [reply]
thanks it worked!

**Comment by MonogameGuy** [comment]
You can use any valid 2D collision method on objects **at the same Z value**.

If free movement is possible, I would use a simple axis aligned `Rectangle` centered and sized to the largest that will fit in one iso-square.

Fixed movement, or free movement against a fixed map, allows for simple and fast grid, array based, collisions.

**Reply by BilalAli2008** [reply]
I tried your method, it did work however in some cases the effect still look 2d
